paddle/framework/tensor.h

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include <memory>
+#include <type_traits>
+#include <typeinfo>
+#include "paddle/framework/ddim.h"
+#include "paddle/platform/assert.h"
+#include "paddle/platform/place.h"
+
+namespace paddle {
+namespace framework {
+
+class Tensor {
+  using paddle::platform::Place;
+  using paddle::platform::get_place;
+
+ public:
+  explicit Tensor(DDim dims) : dims_(dims), place_(get_place()) {}
+  explicit Tensor(DDim dims, Place place) : dims_(dims), place_(place) {}
+
+  Tensor& operator=(const Tensor& src) = delete;
+
+  template <typename T>
+  const T* Data() const {
+    PADDLE_ASSERT(holder_ != nullptr);
+    PADDLE_ASSERT(holder_->Place() == place_);
+    PADDLE_ASSERT(holder_->Size() >= product(dims_) * sizeof(T));
+    return static_cast<const T*>(holder->Ptr());
+  }
+
+  template <typename T>
+  bool NeedReset() const {
+    return (holder_ == nullptr || holder_->Place() != place_ ||
+            holder_->Size() < product(dims_) * sizeof(T));
+  }
+
+  // must be POD types
+  template <typename T, typename = std::enable_if<std::is_pod<T>::value>::type>
+  T* MutableData() {
```

Comment:
> Accessors and mutators (get and set functions) may be named like variables. These often correspond to actual member variables, but this is not required.

Yes, it' ok to using `mutable_data()`.

By the way, do we need to rename functions in `DDim` and `Place` to follow Google C++ style?